### PR TITLE
Robust level 1 validation testing

### PR DESCRIPTION
**Pull request by JRosenkranz**
This PR improves upon shape testing with a more robust level 1 validation testing. This new validation is based on failure rates generated between cpu and cuda for cross-entropy loss and mean diff (per decode step).

### CODE REVIEW

File: aiu_fms_testing_utils/utils/__init__.py [hunk 1]
```diff
@@ -25,4 +31,60 @@ def ids_for_prompt(prompt, tokenizer):
...
+    dataset_path: str,
+    num_requests: int,
+    tokenizer: BaseTokenizer,
+    prompt_length_min: int = 32,
+    prompt_length_max: int = 64,
+    seed: Optional[int] = None
+) -> List[Tuple[str, int]]:
+    if not os.path.exists(dataset_path):
+        print("downloading share-gpt dataset as it does not exist")
+        __download_file("https://huggingface.co/datasets/anon8231489123/ShareGPT_Vicuna_unfiltered/resolve/main/ShareGPT_V3_unfiltered_cleaned_split.json", dataset_path)
```
Comment (ani300) [inline aiu_fms_testing_utils/utils/__init__.py:59]: in case this URL changes, how do we deal with the possible errors on CI?

File: aiu_fms_testing_utils/utils/__init__.py [hunk 2]
```diff
@@ -25,4 +31,60 @@ def ids_for_prompt(prompt, tokenizer):
...
+        if len(filtered_dataset) == num_requests:
+            break
+
+        # Tokenize the prompts and completions.
+        prompt = dataset[i][0]
+        prompt_token_ids = ids_for_prompt(prompt, tokenizer)
+        
+        prompt_len = len(prompt_token_ids)
+        if prompt_len < prompt_length_min or prompt_len > prompt_length_max:
+            # Prune too short sequences.
```
Comment (ani300) [inline aiu_fms_testing_utils/utils/__init__.py:86]: or too long?

File: scripts/generate_metrics.py [hunk 3]
```diff
@@ -231,63 +213,88 @@ def filter_before_eos(l, filter_indexes):
...
+prob_ce_loss_metrics = []
 
-level_1_metrics = capture_level_1_metrics(
-    cpu_validation_info.get_info("logits"),
-    cuda_validation_info.get_info("logits"),
-    top_k_loss_calculator(args.topk_per_token, cross_entropy),
-)
-loss_metrics = filter_before_eos(level_1_metrics, eos_indexes)
-write_csv(loss_metrics, os.path.join(args.output_dir, f"{prefix}.ce.csv"), "ce")
+prefix = f"{args.variant.replace('/', '--')}_max-new-tokens-{args.max_new_tokens}_batch-size-{args.batch_size}_seq-length{args.min_pad_length}_dtype-{args.default_dtype}"
```
Comment (ani300) [inline scripts/generate_metrics.py:237]: add a comment explaining why this naming convention

File: tests/models/test_shapes.py [hunk 4]
```diff
@@ -1,29 +1,55 @@
...
+    failure_rate_threshold = float(failure_rate_threshold)
+
+if isinstance(default_metrics_threshold, str):
+    m_list = [float(m) for m in default_metrics_threshold.split(",")]
+    default_metrics_threshold = (m_list[0], (m_list[1], m_list[2]))
+
+
+# thresholds are chosen based on 1024 tokens per sequence
+# 1% error threshold rate between cpu fp32 and cuda fp16
+# if a models failure thresholds do not exist in this dict, default to the default_metrics_threshold
```
Comment (ani300) [inline tests/models/test_shapes.py:44]: where are the defaults defined?

**Review by ani300 — APPROVED** [review]
left some minor comments on documentation, other than that lgtm, and I guess next step is adding these tests to the CI pipeline?

**Comment by ani300** [comment]
Also DCO is complaining!

File: aiu_fms_testing_utils/utils/__init__.py (hunk 1 above)
Comment (JRosenkranz) [reply aiu_fms_testing_utils/utils/__init__.py:59]: We could download and keep this dataset somewhere, and just set the `SHARE_GPT_DATASET_PATH`. If this dataset is removed, we may need to download from another location. We could also add an environment variable to set the URL to a new URL if need be?
